src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+    super(typeof length === 'object' ? length : { length, decimals });
+  }
+
+  public validate(value: AcceptedNumber): asserts value is AcceptedNumber;
+  public validate(value: AcceptedNumber): true {
+    if (!Validator.isFloat(String(value))) {
+      throw new ValidationError(
+        util.format('%j is not a valid float', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+
+  protected _value(value: AcceptedNumber) {
+    const num = typeof value === 'number' ? value : Number(String(value));
+
+    if (Number.isNaN(num)) {
+      return 'NaN';
+    }
+
+    if (!Number.isFinite(num)) {
+      const sign = num < 0 ? '-' : '';
+
+      return `${sign}Infinity`;
+    }
+
+    return num.toString();
+  }
+
+  protected _stringify(value: AcceptedNumber) {
+    this.validate(value);
+
+    return `'${this._value(value)}'`;
+  }
+
+  protected _bindParam(value: AcceptedNumber, options: BindParamOptions) {
+    return options.bindParam(this._value(value));
+  }
+}
+
+/**
+ * Floating point number (4-byte precision).
+ */
+export type FLOAT = _FLOAT;
+export const FLOAT = classToInvokable(_FLOAT);
+
+class _REAL extends FLOAT {
+  public static readonly key = 'REAL';
+}
+
+/**
+ * Floating point number (4-byte precision).
+ */
+export type REAL = _REAL;
+export const REAL = classToInvokable(_REAL);
+
+class _DOUBLE extends FLOAT {
+  public static readonly key = 'DOUBLE';
+}
+
+/**
+ * Floating point number (8-byte precision).
+ */
+export type DOUBLE = _DOUBLE;
+export const DOUBLE = classToInvokable(_DOUBLE);
+
+interface DecimalOptions extends NumberOptions {
+  scale?: number;
+  precision?: number;
+}
+
+type DecimalConstructorParams =
+  | []
+  | [DecimalOptions]
+  | [number]
+  | [number, number | undefined];
+
+class _DECIMAL extends NUMBER {
+  public static readonly key = 'DECIMAL';
+  protected _scale?: number;
+  protected _precision?: number;
+  protected options: DecimalOptions;
+
+  /**
+   * @param {string|number} [precision] defines precision
+   * @param {string|number} [scale] defines scale
+   */
+  constructor(options?: DecimalOptions);
+  constructor(precision: number, scale?: number);
+  constructor(...args: DecimalConstructorParams);
+  constructor(precision?: number | DecimalOptions, scale?: number) {
+    if (typeof precision === 'object') {
+      super(precision);
+      this.options ??= precision;
+    } else {
+      super();
+      // Sadly TS isn't smart enough to infer that this will be defined by the constructor of NUMBER.
+      // This should never do anything becuase this.options should already be defined.
+      this.options ??= {};
+
+      this.options.precision = precision;
+      this.options.scale = scale;
+    }
+
+    this._precision = this.options.precision;
+    this._scale = this.options.scale;
+  }
+
+  public toSql() {
+    if (this._precision || this._scale) {
+      return `DECIMAL(${[this._precision, this._scale]
```

Comment:
if scale is specified but precision is not, we should throw with a clear error message